File: engine.py
"""
Chaos-based stream cipher engine — RESEARCH ARTIFACT, NOT PRODUCTION CRYPTO.

This implements the user's proposed design faithfully: a discretized, pure-integer
Piecewise Linear Chaotic Map (PWLCM) used as a keystream generator, XOR'd with the
plaintext. Pure-integer math (modulus M = 2**127 - 1, the Mersenne prime M127) makes the
keystream bit-for-bit identical on any CPU/OS — this solves the "finite precision
paradox" that kills floating-point chaos ciphers.

WHAT IS FAITHFUL TO THE ORIGINAL PROPOSAL:
  - PWLCM core map, integer-only.
  - Byte extraction via (state >> 24) & 0xFF.
  - XOR keystream encryption (encrypt == decrypt).

WHAT WAS ADDED so the design can be *fairly tested* (and used without an instant break):
  - A nonce/IV. Keystream = f(key, nonce). Without this, encrypting two messages with
    the same key is a two-time pad — an instant break that has nothing to do with the
    map's quality. attacks/two_time_pad.py deliberately demonstrates that footgun.

SECURITY STATUS: UNVETTED. This is a homemade cipher built to be *attacked and measured*,
not trusted. Do not protect real data with it. See REPORT.md for the empirical verdict.
"""

from __future__ import annotations

# Mersenne prime modulus — the size of the integer "grid" the chaotic state lives on.
# Mersenne prime M127 — the integer "grid" the chaotic state lives on (#1, the bigger grid).
# Moving from 2^61-1 to 2^127-1 raises the per-map orbit length: by the random-function rho law
# the period scales as ~sqrt(M), so this lifts the honest per-map period from ~2^30 to ~2^63.
# (Verify the rho law at small scale with attacks/period_census.py; the 2^63 figure is the law's
# PREDICTION at the new M, not a direct measurement — 2^63 is too large to census.)
M = (1 << 127) - 1         # 170141183460469231731687303715884105727  (Mersenne prime M127)
HALF = M // 2
# Mid-space escape if the map ever hits the 0 fixed point. 128-bit alternating pattern % M lands
# well inside (0, M) for any grid up to 2^128, so this scales with M automatically.
DEAD_STATE_FIX = 0x55555555555555555555555555555555 % M

MASK64 = (1 << 64) - 1

# --- Output hardening (#3 "frosted glass") + throughput (#4) -----------------------------------
# We do NOT emit the raw state. Each chaotic step's state is run through a nonlinear finalizer
# (finalize) and we emit OUTPUT_BYTES_PER_STEP bytes of the result. This knob trades speed vs
# security:
#   * SECURITY: the finalizer (xorshift + multiply, "ARX") destroys the PWLCM's piecewise-LINEAR
#     structure, so an attacker can't algebraically roll the output back to the state — closing the
#     invertibility weakness the old raw "(x >> 24) & 0xFF" left open. We also emit FEWER bytes than
#     the finalized word is wide, so part of every state stays hidden behind the glass.
#   * SPEED: more bytes per expensive step => fewer steps per message (the #4 throughput win).
# The exact value MUST be validated by the Phase-2 attack tooling (correlation / state-recovery),
# never assumed. 4 of 8 bytes is the conservative starting point.
OUTPUT_BYTES_PER_STEP = 4


def finalize(z: int) -> int:
    """Nonlinear ARX 'frosted-glass' output filter (the SplitMix64 / MurmurHash3 fmix64 mixer).

    Maps the 127-bit chaotic state to a well-avalanched 64-bit word with xorshift + multiply. The
    multiplications make it NONLINEAR, so the piecewise-linear PWLCM structure an attacker would use
    to invert the map is destroyed. No division, no tables => fast, Rust-friendly, constant-time.

    HONEST NOTE: this mixer is itself a bijection. The one-wayness an attacker faces comes from
    TRUNCATION (we emit only OUTPUT_BYTES_PER_STEP of the 8 bytes) + the 4-map XOR combiner — not
    from the mix alone. MEASURED in Phase 2 (attacks/differential_attack.py): no avalanche gap (every
    state bit reaches the output ~1/2), no usable single-bit differential, no published->hidden/state
    correlation, and the preimage law ~2^(w/2) holds (=> 2^32 candidates per emitted step, per map)."""
    # Fold the wide state into 64 bits FIRST so every state bit reaches the output. With the bigger
    # grid (#1) the state is up to 127 bits; a bare 64-bit mask would silently drop the top 63 bits.
    # XOR-folding the high half in keeps the whole state live. (No-op for states <= 64 bits.)
    z = (z ^ (z >> 64)) & MASK64
    z = ((z ^ (z >> 30)) * 0xBF58476D1CE4E5B9) & MASK64
    z = ((z ^ (z >> 27)) * 0x94D049BB133111EB) & MASK64
    z ^= z >> 31
    return z


def _kdf_hash(prefix: bytes, master_key: bytes, nonce: bytes, index: int | None = None) -> bytes:
    """SHA-512 over prefix [|| index(2,BE) || '|'] || master_key || '|' || nonce.

    Returns the raw 64-byte digest. Shared across from_master(), multimap._derive_engine(),
    and ctr._block_engine() — one home, no copy-paste."""
    import hashlib

    h = hashlib.sha512(prefix)
    if index is not None:
        h.update(index.to_bytes(2, "big"))
        h.update(b"|")
    h.update(master_key)
    h.update(b"|")
    h.update(nonce)
    return h.digest()


def _derive_seed_control(h: bytes) -> tuple[int, int]:
    """From a 64-byte hash, extract (seed_key: 0..M, control_parameter: 0..HALF)."""
    seed_key = int.from_bytes(h[0:24], "big")
    control = int.from_bytes(h[24:48], "big")
    return seed_key, control


class DiscreteChaoticEngine:
    """Integer PWLCM keystream generator + XOR stream cipher.

    Parameters
    ----------
    seed_key : int
        The secret. Sets the initial chaotic state x0. (>= 122 bits recommended.)
    control_parameter : int
        The map's break-point p (akin to a gravitational constant). Part of the secret.
    nonce : int
        Public, per-message. Must be unique per (key) to avoid keystream reuse.
        Mixed into the initial state so the same key yields a fresh keystream each message.
    """

    # Smallest break-point we allow. p in {0, 1} (and the symmetric top end) put almost
    # the entire state into one branch and collapse the orbit — the period-1 weak class
    # found in adversarial testing. We reject that whole neighbourhood. Defined RELATIVE to the
    # grid (HALF / 2^20) so it scales with M: the period census mirrors exactly this band.
    MIN_P = HALF >> 20

    def __init__(self, seed_key: int, control_parameter: int, nonce: int = 0):
        # --- derive the break-point p, with weak-parameter REJECTION ---
        # p must sit well inside (0, HALF): not tiny, not within MIN_P of HALF.
        self.p = control_parameter % HALF
        if not (self.MIN_P <= self.p <= HALF - self.MIN_P):
            # Map degenerate / weak parameters into the safe middle band instead of
            # collapsing. Deterministic, so sync is preserved.
            self.p = self.MIN_P + (self.p % (HALF - 2 * self.MIN_P))

        # --- derive initial state x from key, then fold the nonce in ---
        # Nonce mixing: a couple of integer-mix rounds so distinct nonces land in
        # very different regions of the state space. Still 100% deterministic.
        x = seed_key % M
        n = nonce % M
        x = (x ^ ((n * 0x9E3779B97F4A7C15) % M)) % M     # golden-ratio odd multiplier mix
        x = (x + ((n << 17) % M) + 1) % M
        # Unconditional avalanche: diffuse ANY (key, nonce) — including all-zero / tiny inputs —
        # across the full 127-bit grid. Without this, a degenerate key with nonce=0 collapses the
        # two mixes above to x = key + 1: a tiny start state that RESONATES with the map into a
        # SHORT CYCLE on the bigger grid (#1) — caught by the period-census edge probe. Two ARX
        # rounds (multiply + xorshift over M) remove that whole failure class. Deterministic, so
        # Alice and Bob still sync. (A good cipher must turn ANY key into a strong initial state.)
        x = (x * 0x9E3779B97F4A7C15) % M
        x ^= x >> 53
        x = (x * 0x2545F4914F6CDD1D + 0x9E3779B97F4A7C15) % M
        x ^= x >> 49
        x %= M
        if x == 0:
            x = DEAD_STATE_FIX
        self.x = x

        # Warm-up: discard initial outputs so the keystream doesn't expose the raw
        # key/nonce-derived state in its first bytes.
        for _ in range(16):
            self._next_state()

        # Output buffer for the multi-byte-per-step (#4) hardened output (#3). Each refill does one
        # chaotic step + one nonlinear finalize and yields OUTPUT_BYTES_PER_STEP bytes.
        self._buf = b""
        self._buf_i = 0

    @classmethod
    def from_master(cls, master_key: bytes, nonce: bytes) -> "DiscreteChaoticEngine":
        """Build an engine from arbitrary byte-string material via a hash-based KDF.

        Any master_key + nonce maps to a well-distributed seed_key and control_parameter,
        so a caller can NEVER accidentally pick a weak key — the hash output is uniform and
        the weak-parameter band is rejected in __init__ anyway. This is how the AEAD layer
        (aead.py) seeds the keystream. Deterministic => Alice and Bob still sync."""
        h = _kdf_hash(b"chaos-pwlcm-v1|seed|", master_key, nonce)
        seed_key, control = _derive_seed_control(h)
        return cls(seed_key, control, nonce=0)         # nonce already mixed into the hash

    def _next_state(self) -> None:
        """One step of the integer PWLCM — BRANCHLESS, constant-time blueprint.

        Pure integer math => bit-identical on all CPUs. This is the constant-time
        rewrite of the original 4-way `if/elif` (kept in git history): instead of
        choosing ONE of the four PWLCM segments based on the secret state (a timing
        leak — different segments could take different time), we compute ALL four
        candidate next-states every step and keep only the one whose region holds,
        via 0/1 masks. Same work regardless of the secret => no branch-timing signal.

        Output is provably identical to the original for every x in [0, M): the four
        region predicates are mutually exclusive and exhaustive over (0, M), and x==0
        / out-of-range falls through to the dead-state escape exactly as before.

        REMAINING TIMING CAVEAT (for the Rust port, not fixable in Python): the two
        divisions below are by the SECRET-derived divisors p and (HALF - p). Hardware
        integer division is data-dependent on many CPUs, so this is a second timing
        leak the branch-removal does NOT close. The Rust core must replace `// p` and
        `// (HALF - p)` with a precomputed-reciprocal multiply-shift (Barrett/Montgomery)
        computed ONCE at key setup, so the per-byte step has no secret-dependent divide.
        Both divisors are always > 0 (p in [MIN_P, HALF - MIN_P]), so every candidate is
        safe to evaluate even in the regions where it is discarded.

        WIDE-MULTIPLY NOTE (#1, the bigger grid): with M = 2^127-1 the products `M * x` are up
        to ~254 bits. Python big-ints absorb this transparently; the Rust core must do a
        128x128 -> 256-bit multiply (or fold it into the Barrett/Montgomery reduction) for each
        of the four candidates. Still constant-time — width is fixed, not secret-dependent.
        """
        x, p = self.x, self.p

        # Four PWLCM segment candidates — all evaluated every step (constant work).
        # Discarded candidates may use "wrong-region" subtractions; harmless, masked out.
        r1 = (M * x) // p                       # region (0, p)
        r2 = (M * (x - p)) // (HALF - p)        # region [p, HALF)
        r3 = (M * (M - p - x)) // (HALF - p)    # region [HALF, M - p)
        r4 = (M * (M - x)) // p                 # region [M - p, M)

        # Region masks (0 or 1). Exactly one is 1 for x in (0, M); all 0 for the dead case.
        in1 = (0 < x) & (x < p)
        in2 = (p <= x) & (x < HALF)
        in3 = (HALF <= x) & (x < (M - p))
        in4 = ((M - p) <= x) & (x < M)
        dead = not (in1 | in2 | in3 | in4)      # x == 0 (the fixed point) or out of range

        # Mask-select: sum of (candidate * its mask). Bit-identical to the if/elif chain.
        self.x = (r1 * in1) + (r2 * in2) + (r3 * in3) + (r4 * in4) + (DEAD_STATE_FIX * dead)

    def generate_byte(self) -> int:
        """Emit one hardened keystream byte. Refills from one chaotic step + the nonlinear
        finalizer (#3) when the buffer empties, taking OUTPUT_BYTES_PER_STEP bytes per step (#4).
        Replaces the old raw "(x >> 24) & 0xFF", which exposed the linear state directly."""
        if self._buf_i >= len(self._buf):
            self._next_state()
            self._buf = finalize(self.x).to_bytes(8, "big")[:OUTPUT_BYTES_PER_STEP]
            self._buf_i = 0
        b = self._buf[self._buf_i]
        self._buf_i += 1
        return b

    def keystream(self, n: int) -> bytes:
        """Return the next n keystream bytes."""
        return bytes(self.generate_byte() for _ in range(n))

    def encrypt(self, data: bytes) -> bytes:
        """XOR data with the keystream. Running the SAME engine state over the output
        decrypts. (Stateful: advances the keystream — make a fresh engine to decrypt.)"""
        out = bytearray(len(data))
        for i, b in enumerate(data):
            out[i] = b ^ self.generate_byte()
        return bytes(out)

    # decrypt is identical to encrypt for an XOR stream cipher
    decrypt = encrypt


def cipher(seed_key: int, control_parameter: int, nonce: int = 0) -> DiscreteChaoticEngine:
    """Convenience factory."""
    return DiscreteChaoticEngine(seed_key, control_parameter, nonce)

File: rust/src/engine.rs
//! Chaos PWLCM keystream core — Rust hot-loop port (Phase 4).
//!
//! This is a faithful, BIT-IDENTICAL re-implementation of the Python `DiscreteChaoticEngine`
//! (engine.py): the same integer PWLCM map on the Mersenne grid M = 2^127 - 1, the same init
//! avalanche, the same nonlinear `finalize` output mixer, the same 4-bytes-per-step buffer.
//!
//! Correctness is PROVEN, not assumed: `cargo test` checks the division invariant and the
//! finalize mixer locally, and the Python test `tests/test_rust_parity.py` checks that this
//! core reproduces every `engine_raw` vector in `kat/vectors.json` byte-for-byte.
//!
//! STAGE B (this file) CLOSES the timing leak. The per-step divide by the secret `p` / `HALF-p`
//! is gone, replaced by a precomputed-reciprocal multiply-shift + one branchless correction. The
//! reciprocal is computed ONCE per key at setup (where a variable-time divide is acceptable — it
//! runs once, not per byte). The hot loop now does only multiplies, shifts, adds, and masked
//! compares — all constant-time on fixed-width limbs.

use ruint::aliases::U256;

// ---- grid constants (mirror engine.py exactly) ----
pub(crate) const M: u128 = (1u128 << 127) - 1; // Mersenne prime M127
pub(crate) const HALF: u128 = M / 2; // 2^126 - 1
pub(crate) const MIN_P: u128 = HALF >> 20;
pub(crate) const DEAD_STATE_FIX: u128 = 0x5555_5555_5555_5555_5555_5555_5555_5555; // < M, so % M is itself

const K_GOLDEN: u128 = 0x9E37_79B9_7F4A_7C15;
const K_MIX: u128 = 0x2545_F491_4F6C_DD1D;
const F_MUL1: u64 = 0xBF58_476D_1CE4_E5B9;
const F_MUL2: u64 = 0x94D0_49BB_1331_11EB;

pub(crate) const OUTPUT_BYTES_PER_STEP: usize = 4;

/// Scale of the precomputed reciprocal: V = floor(M * 2^RECIP_SHIFT / d). 127 gives a truncation
/// error < 1/2 for every divisor in this map (all < 2^126), so a SINGLE branchless correction is exact.
const RECIP_SHIFT: usize = 127;

#[inline]
pub(crate) fn u(v: u128) -> U256 {
    U256::from_limbs([v as u64, (v >> 64) as u64, 0, 0])
}

#[inline]
pub(crate) fn lo128(x: U256) -> u128 {
    let l = x.as_limbs();
    (l[0] as u128) | ((l[1] as u128) << 64)
}

/// Branchless constant-time select: returns `a` if `cond == 1`, else `b`. `cond` must be 0 or 1.
#[inline]
fn select(cond: u128, a: U256, b: U256) -> U256 {
    let m = (cond as u64).wrapping_neg();
    let (la, lb) = (a.as_limbs(), b.as_limbs());
    U256::from_limbs([
        (la[0] & m) | (lb[0] & !m),
        (la[1] & m) | (lb[1] & !m),
        (la[2] & m) | (lb[2] & !m),
        (la[3] & m) | (lb[3] & !m),
    ])
}

/// (a * b) mod M, for a, b < 2^127. Used only at key setup (the init avalanche), not per byte.
#[inline]
fn mulmod_m(a: u128, b: u128) -> u128 {
    lo128((u(a) * u(b)) % u(M))
}

/// The nonlinear "frosted-glass" output mixer (SplitMix64 / fmix64), identical to engine._finalize.
#[inline]
pub(crate) fn finalize(z: u128) -> u64 {
    let mut z = ((z ^ (z >> 64)) & (u64::MAX as u128)) as u64;
    z = (z ^ (z >> 30)).wrapping_mul(F_MUL1);
    z = (z ^ (z >> 27)).wrapping_mul(F_MUL2);
    z ^= z >> 31;
    z
}

/// Precompute the scaled reciprocal V = floor(M * 2^RECIP_SHIFT / d) for a fixed divisor d.
#[inline]
fn reciprocal(d: u128) -> U256 {
    (u(M) << RECIP_SHIFT) / u(d)
}

/// CONSTANT-TIME floor(M * num / den) via the precomputed reciprocal.
///
/// No hardware divide on the secret: q_approx = (num*recip) >> RECIP_SHIFT undershoots the true
/// quotient by at most 1, and one branchless correction makes it exact.
#[inline]
fn div_step(num: u128, den: u128, recip: U256) -> u128 {
    let nu = u(num);
    let q_approx = (nu * recip) >> RECIP_SHIFT;
    let m_num = (nu << 127) - nu;
    let rem = m_num - q_approx * u(den);
    let ge = (rem >= u(den)) as u128;
    lo128(q_approx) + ge
}

/// Big-int reference oracle (Stage A's exact divide). TEST-ONLY.
#[cfg(test)]
fn div_step_oracle(num: u128, den: u128) -> u128 {
    lo128((u(M) * u(num)) / u(den))
}

pub struct ChaosEngine {
    x: u128,
    p: u128,
    recip_p: U256,
    recip_hp: U256,
    buf: [u8; 8],
    buf_len: usize,
    buf_i: usize,
}

impl ChaosEngine {
    /// Mirror of DiscreteChaoticEngine.__init__(seed_key, control_parameter, nonce).
    pub fn new(seed_key: u128, control_parameter: u128, nonce: u128) -> Self {
        let mut p = control_parameter % HALF;
        if !(MIN_P <= p && p <= HALF - MIN_P) {
            p = MIN_P + (p % (HALF - 2 * MIN_P));
        }

        let m256 = u(M);
        let mut x = u(seed_key % M);
        let nn = nonce % M;
        x = (x ^ u(mulmod_m(nn, K_GOLDEN))) % m256;
        x = (x + ((u(nn) << 17) % m256) + U256::from(1u8)) % m256;
        x = (x * u(K_GOLDEN)) % m256;
        x ^= x >> 53;
        x = (x * u(K_MIX) + u(K_GOLDEN)) % m256;
        x ^= x >> 49;
        x %= m256;
        let mut x = lo128(x);
        if x == 0 {
            x = DEAD_STATE_FIX;
        }

        let mut eng = ChaosEngine {
            x,
            p,
            recip_p: reciprocal(p),
            recip_hp: reciprocal(HALF - p),
            buf: [0u8; 8],
            buf_len: 0,
            buf_i: 0,
        };
        for _ in 0..16 {
            eng.next_state();
        }
        eng
    }

    /// One PWLCM step — constant-time, branchless.
    #[inline]
    fn next_state(&mut self) {
        let x = self.x;
        let p = self.p;

        let in1 = ((x > 0) & (x < p)) as u128;
        let in2 = ((x >= p) & (x < HALF)) as u128;
        let in3 = ((x >= HALF) & (x < (M - p))) as u128;
        let in4 = ((x >= (M - p)) & (x < M)) as u128;
        let dead = 1u128 - (in1 | in2 | in3 | in4);

        let n1 = x;
        let n2 = x.wrapping_sub(p);
        let n3 = M.wrapping_sub(p).wrapping_sub(x);
        let n4 = M.wrapping_sub(x);
        let num = n1
            .wrapping_mul(in1)
            .wrapping_add(n2.wrapping_mul(in2))
            .wrapping_add(n3.wrapping_mul(in3))
            .wrapping_add(n4.wrapping_mul(in4));

        let sel_p = in1 | in4;
        let sel_hp = in2 | in3;
        let den = p
            .wrapping_mul(sel_p)
            .wrapping_add((HALF - p).wrapping_mul(sel_hp))
            .wrapping_add(dead);

        let recip = select(sel_p, self.recip_p, select(sel_hp, self.recip_hp, U256::ZERO));
        let q = div_step(num, den, recip);
        self.x = q.wrapping_mul(1 - dead).wrapping_add(DEAD_STATE_FIX.wrapping_mul(dead));
    }

    /// Step once and return this step's OUTPUT_BYTES_PER_STEP keystream bytes.
    #[inline]
    pub fn next_block(&mut self) -> [u8; OUTPUT_BYTES_PER_STEP] {
        self.next_state();
        let full = finalize(self.x).to_be_bytes();
        let mut out = [0u8; OUTPUT_BYTES_PER_STEP];
        out.copy_from_slice(&full[..OUTPUT_BYTES_PER_STEP]);
        out
    }

    #[inline]
    fn refill(&mut self) {
        let block = self.next_block();
        self.buf[..OUTPUT_BYTES_PER_STEP].copy_from_slice(&block);
        self.buf_len = OUTPUT_BYTES_PER_STEP;
        self.buf_i = 0;
    }

    #[inline]
    pub fn next_byte(&mut self) -> u8 {
        if self.buf_i >= self.buf_len {
            self.refill();
        }
        let b = self.buf[self.buf_i];
        self.buf_i += 1;
        b
    }

    pub fn keystream(&mut self, n: usize) -> Vec<u8> {
        let mut out = Vec::with_capacity(n);
        for _ in 0..n {
            out.push(self.next_byte());
        }
        out
    }

    pub fn encrypt(&mut self, data: &[u8]) -> Vec<u8> {
        let ks = self.keystream(data.len());
        data.iter().zip(ks.iter()).map(|(p, k)| p ^ k).collect()
    }

    /// KDF constructor — mirror of engine.py `DiscreteChaoticEngine.from_master`.
    pub fn from_master(master_key: &[u8], nonce: &[u8]) -> Self {
        let h = crate::utils::kdf_hash(b"chaos-pwlcm-v1|seed|", None, master_key, nonce);
        let (seed, control) = crate::utils::derive_seed_control(&h);
        ChaosEngine::new(seed, control, 0)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn finalize_matches_known_answer() {
        assert_eq!(finalize(1), 0x5692_161d_100b_05e5);
        assert_eq!(finalize(0), 0);
    }

    #[test]
    fn div_step_invariant_holds() {
        let cases: [(u128, u128); 7] = [
            (12345, MIN_P + 12345),
            (HALF - 1, HALF),
            (MIN_P, MIN_P + 7),
            (1, HALF - MIN_P),
            (0x0123_4567_89AB_CDEF, 0x0FED_CBA9_8765_4321),
            (M / 5, M / 3),
            (MIN_P + 1, MIN_P),
        ];
        for (num, den) in cases {
            let q = div_step(num, den, reciprocal(den));
            let n = u(M) * u(num);
            let lhs = u(q) * u(den);
            let rhs = u(q.wrapping_add(1)) * u(den);
            assert!(lhs <= n, "q*den <= M*num failed for ({num},{den})");
            assert!(n < rhs, "M*num < (q+1)*den failed for ({num},{den})");
        }
    }

    struct Rng(u64, u64);
    impl Rng {
        fn next(&mut self) -> u64 {
            let mut s1 = self.0;
            let s0 = self.1;
            self.0 = s0;
            s1 ^= s1 << 23;
            self.1 = s1 ^ s0 ^ (s1 >> 17) ^ (s0 >> 26);
            self.1.wrapping_add(s0)
        }
        fn u128(&mut self) -> u128 {
            ((self.next() as u128) << 64) | (self.next() as u128)
        }
    }

    #[test]
    fn recip_div_matches_bigint_oracle() {
        let mut rng = Rng(0x1234_5678_9abc_def0, 0x0fed_cba9_8765_4321);
        let span = HALF - 2 * MIN_P;
        let divisors = if cfg!(debug_assertions) { 60 } else { 400 };
        let nums = if cfg!(debug_assertions) { 4_000 } else { 12_000 };
        for di in 0..divisors {
            let den = match di {
                0 => MIN_P,
                1 => HALF - MIN_P,
                2 => HALF / 2,
                _ => MIN_P + (rng.u128() % span),
            };
            let recip = reciprocal(den);
            for &num in &[0u128, 1, den - 1, den, den + 1] {
                assert_eq!(
                    div_step(num, den, recip),
                    div_step_oracle(num, den),
                    "edge num={num} den={den}"
                );
            }
            for _ in 0..nums {
                let num = rng.u128() % (den + 2);
                assert_eq!(
                    div_step(num, den, recip),
                    div_step_oracle(num, den),
                    "num={num} den={den}"
                );
            }
        }
    }

    #[test]
    fn keystream_is_deterministic() {
        let a = ChaosEngine::new(11, 22, 33).keystream(64);
        let b = ChaosEngine::new(11, 22, 33).keystream(64);
        assert_eq!(a, b);
    }
}
